### PR TITLE
feat: wrappers for vscode multi cursor commands

### PR DESCRIPTION
**Pull request by xiyaowong**
- [x] `editor.action.addSelectionToNextFindMatch`
- [x] `editor.action.addSelectionToPreviousFindMatch`
- [x] `editor.action.selectHighlights`
- [x] `selectAllSearchEditorMatches`

Example:
Add selection to next find match:
- Invoke action in nvim: `vscode.action('editor.action.addSelectionToNextFindMatch')`
- Execute command in vscode: `neovim:editor.action.addSelectionToNextFindMatch`

keybindings.json
```json
  {
    "command": "neovim:editor.action.addSelectionToNextFindMatch",
    "key": "ctrl+d",
    "when": "editorFocus"
  }
```

Resolve: #1145

### CODE REVIEW

**Comment by xiyaowong** [comment]
@archilkarchava The approach is the same, but it is more convenient and concise in vscode, and it is easier to handle edge cases, such as cursor position jitter.

**Comment by theol0403** [comment]
I think it would be good to make this more general. We should provide a generic "execute this command from normal mode that moves into insert mode and creates a selection.

Then, the API could be (rough example):

```
{
command: "neovim:wrap-cursor-command",
args: "addNextSelection"
}
```

**Comment by theol0403** [comment]
The reason is that there are a lot of multi cursor commands that would be good to wrap without maintenance burden.

File: src/cursor_manager.ts [hunk 1]
```diff
@@ -528,6 +534,58 @@ export class CursorManager implements Disposable {
...
+
+        if (this.main.modeManager.isInsertMode) {
+            await run();
+            return;
+        }
+
+        if (editor.selection.isEmpty) {
+            await run();
+            const selections = editor.selections;
+            this.main.client.input("<Esc>"); // Correcting the position
```
Comment (theol0403) [inline src/cursor_manager.ts:552]: I don't love the delays. Maybe you can await the input and then await a cursor update.

**Comment by xiyaowong** [comment]
> I think it would be good to make this more general. We should provide a generic "execute this command from normal mode that moves into insert mode and creates a selection.
> 
> Then, the API could be (rough example):
> 
> ```
> {
> command: "neovim:wrap-cursor-command",
> args: "addNextSelection"
> }
> ```

I don't think this is better. With prefix "neovim:", the user can understand at a glance what the command is for. Where the maintenance cost is increased, the command is automatically registered and unified.

**Comment by xiyaowong** [comment]
Even if users don't know these functions, users can directly use `editor.action.addSelectionToNextFindMatch`, which is automatically replaced, which is also the design intention of Action in PR lua api.

File: src/cursor_manager.ts (hunk 1 above)
Comment (xiyaowong) [reply src/cursor_manager.ts:519]: > await the input

I just discovered that "input" can be awaited. :smile:

**Comment by theol0403** [comment]
What if you want to wrap `editor.action.smartSelect.grow` (random example), or maybe a command from another vscode extension. You would have to make a PR here every time. Is there not a clean way to make a generic wrapper for any vscode command that you want to transition into insert mode?

**Comment by xiyaowong** [comment]
These wrappers themselves are special cases, and the few built-in commands are enough.

**Comment by xiyaowong** [comment]
This is a general example in VSCode, hoping to be helpful to those in need

```lua
vim.keymap.set({ 'n', 'i', 'x' }, '<C-d>', function()
  local vscode = require('vscode-neovim')
  local mode = vim.api.nvim_get_mode().mode

  local select = function(cb)
    vscode.action('editor.action.addSelectionToNextFindMatch', { callback = cb })
  end

  if mode == 'i' then return select() end

  if mode:lower() == 'v' then
    vim.api.nvim_input('<ESC>a')
    vim.schedule(function()
      local range = vim.lsp.util.make_given_range_params(nil, nil, 0, 'utf-16').range
      vscode.action('noop', {
        range = range,
        callback = select,
        restore_selection = false,
      })
    end)
    return
  end

  if mode == 'n' then select(function()
    vim.api.nvim_input('<ESC>')
    select()
  end) end
end)
```

**Comment by theol0403** [comment]
> This is a general example in nvim, hoping to be helpful to those in need

Again, is there any way we can provide this as part of the API or included as a helper function?

File: src/cursor_manager.ts [hunk 2]
```diff
@@ -495,6 +501,57 @@ export class CursorManager implements Disposable {
...
+
+        await cursorReady();
+
+        if (this.main.modeManager.isInsertMode) return run();
+
+        if (editor.selection.isEmpty) {
+            await run();
+            const selections = editor.selections;
+            await this.main.client.input("<Esc>"); // Correcting the position
+            await wait(100); // Compensate for debounce time
```
Comment (theol0403) [inline src/cursor_manager.ts:519]: Can you pull this constant in, instead of magic number?

**Comment by xiyaowong** [comment]
> > This is a general example in nvim, hoping to be helpful to those in need
> 
> Again, is there any way we can provide this as part of the API or included as a helper function?

There should be, but there should be more things to consider in implementation. 

Provides a simple action to switch from normal/visual to insert mode, but keeps the selections of vscode unchanged.

**Comment by xiyaowong** [comment]
These commands can be implemented in nvim. All crowded into the project will really make the project bloated. Closed this pr.

**Comment by xiyaowong** [comment]
@theol0403 
It seems like I made several serious mistakes. I tried to click the sync button in vscode as usual, but I found out that it actually pushed the commits to the master branch. Is there any way to revert these commits? I am truly sorry. 😭 

image

**Comment by theol0403** [comment]
@xiyaowong you can revert them manually, or make a force push as a last resort. You may have to temporarily disable branch protection rules to allow that.

**Pull request closed without merging**